### ISSUE TITLE
faster DOM renderer

### ISSUE THREAD

**Issue by jerch** [xtermjs/xterm.js#4604]
The performance of the DOM renderer is very painful. In several local attempts I tried to isolate the reasons behind this and well, it mainly boils down to the sheer amount of [span elements x CSS rules], which gives the browser engine a hard time to update & layout things.

I was able to find a solution to reduce the "element pressure" by merging cells, that share the same fg/bg settings and where the char content gets properly layouted by the browsers font renderer. This gives a very nice perf boost (test is `ls -lR /usr`):

- master
  ![grafik](https://github.com/xtermjs/xterm.js/assets/6193135/cddc443f-fc88-447c-afa6-091205a0ad8f)
- WIP branch
  ![grafik](https://github.com/xtermjs/xterm.js/assets/6193135/865c0de5-8d0b-4253-8471-3a5ea7d28427)
- compared to webgl renderer
  ![grafik](https://github.com/xtermjs/xterm.js/assets/6193135/31f0ed54-e2d3-4e05-b745-3684d7f137ef)

Observations from above:
- no more red flagged frames in devtools - most frames can be updated in reasonable time and we see almost no discarded screen refreshs anymore
- the "stattering" got much better - while the DOM renderer still saturates the CPU (the pie chart is still at ~100% usage), the rendering task is like ~7 times faster lifting some of the burden
- JS runtime got much worse, prolly due to the additional merging logic (well, its just a quickly hacked WIP atm)
- overall runtime is on par with webgl renderer with ~14s (prolly hitting an OS/PTY limit here)
- ofc webgl is still far ahead finishing the test rather quick while the CPU idles for ~40% (scales much better in bigger integrations)

So while this looks rather promising, there are ofc downsides - most prominent the fact, that a cell does not strictly refer to a single span element anymore. This currently creates glitches during selection, and most likely will break with any other 3rd party code relying on the a nice cell-span matrix (already an issue for double width chars). Furthermore the merging relies on a "close-enough" threshold approximation, thus will create tiny pixel offsets for bigger merges within a line.

**Comment by jerch** [comment]
Random sidenote on these ~14s overall runtime:

I wondered how other TEs would perform, so here are the numbers (all "my machine"-TM, ordered by runtime):

| TE  | runtime in s  |
|---|---|
| alacritty  | 9.3  |
| urxvt  | 9.6  |
| contour  | 9.9  |
| kitty  | 10.1  |
| mlterm  | 10.2  |
| gnome-terminal  | 13.4  |
| xtermjs  | 14.1 |
| xterm  | 20.5  |

So yeah - our demo has ~50% runtime overhead to the fastest. But where does it come from? So I disabled all render parts and measured again - now yielding ~13.4s, with 4.2s Javascript runtime for parser and buffer handling. Means ~9s wasted waiting for data. :see_no_evil: 
Long story short - the overhead mainly results from our "long line" to get the data into the browser realm (node-pty --> websocket --> browser), thus it is a transport artefact. At least the webgl renderer has enough free resources to play in the same ballpark as the fast TEs, if the data could be delivered faster. The DOM renderer would still choke on that.

Overall it is still remarkable that JS with its single context burden allows this execution speed. So kudos to all JIT devs.
